File: Java/sorting/MergeSort.java
/**
 * Merge Sort Algorithm
 * 
 * Time Complexity: O(n log n) in all cases (worst, average, and best)
 * Space Complexity: O(n)
 * 
 * Merge sort is a divide-and-conquer algorithm that divides the input array
 * into two halves,
 * recursively sorts them, and then merges the two sorted halves.
 * It is a stable sorting algorithm and guarantees O(n log n) time complexity.
 */
public class MergeSort {

  /**
   * Sorts an array using merge sort algorithm (ascending order).
   * 
   * @param arr Array of integers to be sorted
   * @return Sorted array
   */
  public static int[] mergeSort(int[] arr) {
    if (arr.length <= 1) {
      return arr;
    }

    mergeSortHelper(arr, 0, arr.length - 1);
    return arr;
  }

  /**
   * Helper method for merge sort that recursively divides the array.
   * 
   * @param arr   Array to be sorted
   * @param left  Starting index
   * @param right Ending index
   */
  private static void mergeSortHelper(int[] arr, int left, int right) {
    if (left < right) {
      // Find the middle point
      int mid = left + (right - left) / 2;

      // Sort first and second halves
      mergeSortHelper(arr, left, mid);
      mergeSortHelper(arr, mid + 1, right);

      // Merge the sorted halves
      merge(arr, left, mid, right);
    }
  }

  /**
   * Merges two sorted subarrays of arr[].
   * First subarray is arr[left..mid]
   * Second subarray is arr[mid+1..right]
   * 
   * @param arr   Array containing the subarrays
   * @param left  Starting index of first subarray
   * @param mid   Ending index of first subarray
   * @param right Ending index of second subarray
   */
  private static void merge(int[] arr, int left, int mid, int right) {
    // Find sizes of two subarrays to be merged
    int n1 = mid - left + 1;
    int n2 = right - mid;

    // Create temporary arrays
    int[] leftArray = new int[n1];
    int[] rightArray = new int[n2];

    // Copy data to temporary arrays
    for (int i = 0; i < n1; i++) {
      leftArray[i] = arr[left + i];
    }
    for (int j = 0; j < n2; j++) {
      rightArray[j] = arr[mid + 1 + j];
    }

    // Merge the temporary arrays back into arr[left..right]
    int i = 0, j = 0;
    int k = left;

    while (i < n1 && j < n2) {
      if (leftArray[i] <= rightArray[j]) {
        arr[k] = leftArray[i];
        i++;
      } else {
        arr[k] = rightArray[j];
        j++;
      }
      k++;
    }

    // Copy remaining elements of leftArray[] if any
    while (i < n1) {
      arr[k] = leftArray[i];
      i++;
      k++;
    }

    // Copy remaining elements of rightArray[] if any
    while (j < n2) {
      arr[k] = rightArray[j];
      j++;
      k++;
    }
  }

  /**
   * Sorts an array in descending order using merge sort.
   * 
   * @param arr Array of integers to be sorted
   * @return Sorted array in descending order
   */
  public static int[] mergeSortDescending(int[] arr) {
    if (arr.length <= 1) {
      return arr;
    }

    mergeSortDescendingHelper(arr, 0, arr.length - 1);
    return arr;
  }

  /**
   * Helper method for descending merge sort.
   */
  private static void mergeSortDescendingHelper(int[] arr, int left, int right) {
    if (left < right) {
      int mid = left + (right - left) / 2;

      mergeSortDescendingHelper(arr, left, mid);
      mergeSortDescendingHelper(arr, mid + 1, right);

      mergeDescending(arr, left, mid, right);
    }
  }

  /**
   * Merges two subarrays in descending order.
   */
  private static void mergeDescending(int[] arr, int left, int mid, int right) {
    int n1 = mid - left + 1;
    int n2 = right - mid;

    int[] leftArray = new int[n1];
    int[] rightArray = new int[n2];

    for (int i = 0; i < n1; i++) {
      leftArray[i] = arr[left + i];
    }
    for (int j = 0; j < n2; j++) {
      rightArray[j] = arr[mid + 1 + j];
    }

    int i = 0, j = 0;
    int k = left;

    while (i < n1 && j < n2) {
      if (leftArray[i] >= rightArray[j]) {
        arr[k] = leftArray[i];
        i++;
      } else {
        arr[k] = rightArray[j];
        j++;
      }
      k++;
    }

    while (i < n1) {
      arr[k] = leftArray[i];
      i++;
      k++;
    }

    while (j < n2) {
      arr[k] = rightArray[j];
      j++;
      k++;
    }
  }

  /**
   * Utility method to print an array
   */
  public static void printArray(int[] arr) {
    for (int i = 0; i < arr.length; i++) {
      System.out.print(arr[i] + " ");
    }
    System.out.println();
  }

  /**
   * Main method with example usage
   */
  public static void main(String[] args) {
    System.out.println("Merge Sort Examples:");

    // Test ascending sort
    int[] arr1 = { 64, 34, 25, 12, 22, 11, 90 };
    System.out.print("\nOriginal array: ");
    printArray(arr1);

    mergeSort(arr1);
    System.out.print("Sorted array (ascending): ");
    printArray(arr1);

    // Test descending sort
    int[] arr2 = { 64, 34, 25, 12, 22, 11, 90 };
    mergeSortDescending(arr2);
    System.out.print("Sorted array (descending): ");
    printArray(arr2);

    // Test with already sorted array
    int[] arr3 = { 1, 2, 3, 4, 5 };
    System.out.print("\nAlready sorted array: ");
    printArray(arr3);
    mergeSort(arr3);
    System.out.print("After merge sort: ");
    printArray(arr3);

    // Test with single element
    int[] arr4 = { 42 };
    System.out.print("\nSingle element array: ");
    printArray(arr4);
    mergeSort(arr4);
    System.out.print("After merge sort: ");
    printArray(arr4);

    // Test with large array
    int[] arr5 = { 38, 27, 43, 3, 9, 82, 10 };
    System.out.print("\nLarge array: ");
    printArray(arr5);
    mergeSort(arr5);
    System.out.print("After merge sort: ");
    printArray(arr5);
  }
}

File: C/sorting/merge_sort.c
/*
 * File: merge_sort.c
 * Description: Implementation of Merge Sort algorithm in C
 * Author: Gompa Rani Prasanna
 *
 * Algorithm:
 *   - Merge Sort is a Divide and Conquer algorithm.
 *   - It divides the array into two halves, recursively sorts them,
 *     and then merges the two sorted halves.
 *
 * Steps:
 *   1. Divide the array into two halves.
 *   2. Recursively sort both halves.
 *   3. Merge the two sorted halves into one sorted array.
 *
 * Time Complexity:
 *   - Best Case: O(n log n)
 *   - Average Case: O(n log n)
 *   - Worst Case: O(n log n)
 *
 * Space Complexity: O(n) (for the temporary arrays used during merging)
 *
 * Example Usage:
 *   Input: arr = [38, 27, 43, 3, 9, 82, 10]
 *   Output: [3, 9, 10, 27, 38, 43, 82]
 */

#include <stdio.h>
#include <stdlib.h>

// Function to merge two sorted subarrays
// arr[l..m] and arr[m+1..r]
void merge(int arr[], int l, int m, int r) {
    int n1 = m - l + 1;
    int n2 = r - m;

    // Create temporary arrays
    int* L = (int*)malloc(n1 * sizeof(int));
    int* R = (int*)malloc(n2 * sizeof(int));

    // Copy data to temp arrays L[] and R[]
    for (int i = 0; i < n1; i++)
        L[i] = arr[l + i];
    for (int j = 0; j < n2; j++)
        R[j] = arr[m + 1 + j];

    // Merge the temp arrays back into arr[l..r]
    int i = 0, j = 0, k = l;
    while (i < n1 && j < n2) {
        if (L[i] <= R[j])
            arr[k++] = L[i++];
        else
            arr[k++] = R[j++];
    }

    // Copy remaining elements of L[], if any
    while (i < n1)
        arr[k++] = L[i++];

    // Copy remaining elements of R[], if any
    while (j < n2)
        arr[k++] = R[j++];

    // Free temporary arrays
    free(L);
    free(R);
}

// Recursive function to perform merge sort
void mergeSort(int arr[], int l, int r) {
    if (l < r) {
        int m = l + (r - l) / 2; // same as (l + r)/2 but avoids overflow
        mergeSort(arr, l, m);    // sort first half
        mergeSort(arr, m + 1, r); // sort second half
        merge(arr, l, m, r);     // merge the two halves
    }
}

// Function to print array
void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

// Example usage
int main() {
    int arr[] = {38, 27, 43, 3, 9, 82, 10};
    int n = sizeof(arr) / sizeof(arr[0]);

    printf("Original array:\n");
    printArray(arr, n);

    mergeSort(arr, 0, n - 1);

    printf("\nSorted array:\n");
    printArray(arr, n);

    return 0;
}
